File: crates/presentation/src/sas.rs
//! SAS (Short Authentication String) generation
//!
//! Generates a human-comparable authentication string for device verification.
//! Uses BLAKE3 hash of public keys and client nonce, converted to emojis.

use blake3::Hasher;

/// SAS emoji list (256 emojis for 8-bit encoding per emoji)
/// Using Signal-style emoji set for familiarity
/// IMPORTANT: This table MUST match the frontend SAS_EMOJIS in web/src/shared/lib/crypto/sas.ts
const SAS_EMOJIS: &[&str] = &[
    // Animals (0-63)
    "🐶", "🐱", "🐭", "🐹", "🐰", "🦊", "🐻", "🐼",
    "🐨", "🐯", "🦁", "🐮", "🐷", "🐸", "🐵", "🐔",
    "🐧", "🐦", "🐤", "🦆", "🦅", "🦉", "🦇", "🐺",
    "🐗", "🐴", "🦄", "🐝", "🐛", "🦋", "🐌", "🐞",
    "🐜", "🦟", "🦗", "🦂", "🐢", "🐍", "🦎", "🦖",
    "🦕", "🐙", "🦑", "🦐", "🦞", "🦀", "🐡", "🐠",
    "🐟", "🐬", "🐳", "🐋", "🦈", "🐊", "🐅", "🐆",
    "🦓", "🦍", "🦧", "🐘", "🦛", "🦏", "🐪", "🐫",
    // Food (64-127)
    "🍎", "🍐", "🍊", "🍋", "🍌", "🍉", "🍇", "🍓",
    "🫐", "🍈", "🍒", "🍑", "🥭", "🍍", "🥥", "🥝",
    "🍅", "🍆", "🥑", "🥦", "🥬", "🥒", "🌶️", "🫑",
    "🌽", "🥕", "🧄", "🧅", "🥔", "🍠", "🥐", "🥯",
    "🍞", "🥖", "🥨", "🧀", "🥚", "🍳", "🧈", "🥞",
    "🧇", "🥓", "🥩", "🍗", "🍖", "🦴", "🌭", "🍔",
    "🍟", "🍕", "🫓", "🥪", "🥙", "🧆", "🌮", "🌯",
    "🫔", "🥗", "🥘", "🫕", "🍝", "🍜", "🍲", "🍛",
    // Nature & Plants (128-191)
    "🌲", "🌳", "🌴", "🌵", "🌾", "🌿", "☘️", "🍀",
    "🍁", "🍂", "🍃", "🍄", "🌸", "💐", "🌷", "🌹",
    "🥀", "🌺", "🌻", "🌼", "🌱", "🪴", "🪵", "🪨",
    "⛰️", "🏔️", "🌋", "🗻", "🏕️", "🏖️", "🏜️", "🏝️",
    "🌅", "🌄", "🌠", "🎇", "🎆", "🌈", "☀️", "🌤️",
    "⛅", "🌦️", "🌧️", "⛈️", "🌩️", "🌨️", "❄️", "☃️",
    "⛄", "🌬️", "💨", "🌪️", "🌫️", "🌊", "💧", "💦",
    "☔", "🌙", "⭐", "🌟", "✨", "💫", "🌍", "🌎",
    // Sports & Activities (192-223)
    "⚽", "🏀", "🏈", "⚾", "🥎", "🎾", "🏐", "🏉",
    "🥏", "🎱", "🪀", "🏓", "🏸", "🏒", "🏑", "🥍",
    "🏏", "🪃", "🥅", "⛳", "🪁", "🏹", "🎣", "🤿",
    "🥊", "🥋", "🎽", "🛹", "🛼", "🛷", "⛸️", "🥌",
    // Objects & Symbols (224-255)
    "🎸", "🪕", "🎹", "🎺", "🎻", "🪘", "🥁", "🪗",
    "🎤", "🎧", "📻", "🎬", "🎨", "🎭", "🎪", "🎰",
    "🚗", "🚕", "🚙", "🚌", "🚎", "🏎️", "🚓", "🚑",
    "🚒", "🚐", "🛻", "🚚", "🚛", "🚜", "✈️", "🚀",
];

/// Generate SAS from public keys and nonce
///
/// ADR-008 compliant: Uses BLAKE3 hash of:
/// - Identity signing public key
/// - Device signing public key
/// - Device ECDH public key
/// - Client nonce (16 bytes)
///
/// Returns 7 emojis representing the first 56 bits of the hash
pub fn generate_sas(
    identity_signing_pk: &[u8],
    device_signing_pk: &[u8],
    device_ecdh_pk: &[u8],
    client_nonce: &[u8],
) -> String {
    let mut hasher = Hasher::new();
    hasher.update(identity_signing_pk);
    hasher.update(device_signing_pk);
    hasher.update(device_ecdh_pk);
    hasher.update(client_nonce);

    let hash = hasher.finalize();
    let bytes = hash.as_bytes();

    // Convert first 7 bytes to 7 emojis (one byte per emoji)
    let mut emojis = Vec::with_capacity(7);
    for &byte in &bytes[..7] {
        emojis.push(SAS_EMOJIS[byte as usize]);
    }

    emojis.join("")
}

/// Generate SAS as a list of emoji indices (for API response)
pub fn generate_sas_indices(
    identity_signing_pk: &[u8],
    device_signing_pk: &[u8],
    device_ecdh_pk: &[u8],
    client_nonce: &[u8],
) -> Vec<u8> {
    let mut hasher = Hasher::new();
    hasher.update(identity_signing_pk);
    hasher.update(device_signing_pk);
    hasher.update(device_ecdh_pk);
    hasher.update(client_nonce);

    let hash = hasher.finalize();
    hash.as_bytes()[..7].to_vec()
}

/// Get the emoji list for client-side rendering
pub fn get_emoji_list() -> Vec<&'static str> {
    SAS_EMOJIS.to_vec()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_sas_deterministic() {
        let identity_pk = [1u8; 32];
        let device_signing_pk = [2u8; 32];
        let device_ecdh_pk = [3u8; 32];
        let nonce = [4u8; 16];

        let sas1 = generate_sas(&identity_pk, &device_signing_pk, &device_ecdh_pk, &nonce);
        let sas2 = generate_sas(&identity_pk, &device_signing_pk, &device_ecdh_pk, &nonce);

        assert_eq!(sas1, sas2);
        // Verify 7 indices are produced (chars().count() is unreliable for multi-codepoint emojis)
        let indices = generate_sas_indices(&identity_pk, &device_signing_pk, &device_ecdh_pk, &nonce);
        assert_eq!(indices.len(), 7);
    }

    #[test]
    fn test_generate_sas_changes_with_input() {
        let identity_pk = [1u8; 32];
        let device_signing_pk = [2u8; 32];
        let device_ecdh_pk = [3u8; 32];
        let nonce1 = [4u8; 16];
        let nonce2 = [5u8; 16];

        let sas1 = generate_sas(&identity_pk, &device_signing_pk, &device_ecdh_pk, &nonce1);
        let sas2 = generate_sas(&identity_pk, &device_signing_pk, &device_ecdh_pk, &nonce2);

        assert_ne!(sas1, sas2);
    }

    #[test]
    fn test_emoji_list_size() {
        assert_eq!(SAS_EMOJIS.len(), 256);
    }

    #[test]
    fn test_no_duplicate_emojis() {
        let mut seen = std::collections::HashSet::new();
        for (i, emoji) in SAS_EMOJIS.iter().enumerate() {
            assert!(seen.insert(*emoji), "Duplicate emoji at index {i}: {emoji}");
        }
    }

    #[test]
    fn test_emoji_table_spot_check() {
        // Spot-check key positions to guard against frontend/backend drift
        // These values must match web/src/shared/lib/crypto/sas.ts
        assert_eq!(SAS_EMOJIS[0], "🐶");
        assert_eq!(SAS_EMOJIS[63], "🐫");
        assert_eq!(SAS_EMOJIS[64], "🍎");
        assert_eq!(SAS_EMOJIS[127], "🍛");
        assert_eq!(SAS_EMOJIS[128], "🌲");
        assert_eq!(SAS_EMOJIS[191], "🌎");
        assert_eq!(SAS_EMOJIS[192], "⚽");
        assert_eq!(SAS_EMOJIS[223], "🥌");
        assert_eq!(SAS_EMOJIS[224], "🎸");
        assert_eq!(SAS_EMOJIS[255], "🚀");
    }
}

File: web/src/shared/lib/crypto/sas.ts
/**
 * SAS (Short Authentication String) Generation
 *
 * ADR-008 compliant: BLAKE3 hash → 7 emojis (56 bits)
 * Used for device verification during multi-device registration.
 */

import { blake3 } from '@noble/hashes/blake3.js'

/**
 * 256 emoji set for SAS display
 * Each byte (0-255) maps to one emoji
 * Selected from Signal-style categories: animals, food, nature, sports, objects
 */
export const SAS_EMOJIS: readonly string[] = [
  // Animals (0-63)
  '🐶', '🐱', '🐭', '🐹', '🐰', '🦊', '🐻', '🐼',
  '🐨', '🐯', '🦁', '🐮', '🐷', '🐸', '🐵', '🐔',
  '🐧', '🐦', '🐤', '🦆', '🦅', '🦉', '🦇', '🐺',
  '🐗', '🐴', '🦄', '🐝', '🐛', '🦋', '🐌', '🐞',
  '🐜', '🦟', '🦗', '🦂', '🐢', '🐍', '🦎', '🦖',
  '🦕', '🐙', '🦑', '🦐', '🦞', '🦀', '🐡', '🐠',
  '🐟', '🐬', '🐳', '🐋', '🦈', '🐊', '🐅', '🐆',
  '🦓', '🦍', '🦧', '🐘', '🦛', '🦏', '🐪', '🐫',

  // Food (64-127)
  '🍎', '🍐', '🍊', '🍋', '🍌', '🍉', '🍇', '🍓',
  '🫐', '🍈', '🍒', '🍑', '🥭', '🍍', '🥥', '🥝',
  '🍅', '🍆', '🥑', '🥦', '🥬', '🥒', '🌶️', '🫑',
  '🌽', '🥕', '🧄', '🧅', '🥔', '🍠', '🥐', '🥯',
  '🍞', '🥖', '🥨', '🧀', '🥚', '🍳', '🧈', '🥞',
  '🧇', '🥓', '🥩', '🍗', '🍖', '🦴', '🌭', '🍔',
  '🍟', '🍕', '🫓', '🥪', '🥙', '🧆', '🌮', '🌯',
  '🫔', '🥗', '🥘', '🫕', '🍝', '🍜', '🍲', '🍛',

  // Nature & Plants (128-191)
  '🌲', '🌳', '🌴', '🌵', '🌾', '🌿', '☘️', '🍀',
  '🍁', '🍂', '🍃', '🍄', '🌸', '💐', '🌷', '🌹',
  '🥀', '🌺', '🌻', '🌼', '🌱', '🪴', '🪵', '🪨',
  '⛰️', '🏔️', '🌋', '🗻', '🏕️', '🏖️', '🏜️', '🏝️',
  '🌅', '🌄', '🌠', '🎇', '🎆', '🌈', '☀️', '🌤️',
  '⛅', '🌦️', '🌧️', '⛈️', '🌩️', '🌨️', '❄️', '☃️',
  '⛄', '🌬️', '💨', '🌪️', '🌫️', '🌊', '💧', '💦',
  '☔', '🌙', '⭐', '🌟', '✨', '💫', '🌍', '🌎',

  // Sports & Activities (192-223)
  '⚽', '🏀', '🏈', '⚾', '🥎', '🎾', '🏐', '🏉',
  '🥏', '🎱', '🪀', '🏓', '🏸', '🏒', '🏑', '🥍',
  '🏏', '🪃', '🥅', '⛳', '🪁', '🏹', '🎣', '🤿',
  '🥊', '🥋', '🎽', '🛹', '🛼', '🛷', '⛸️', '🥌',

  // Objects & Symbols (224-255)
  '🎸', '🪕', '🎹', '🎺', '🎻', '🪘', '🥁', '🪗',
  '🎤', '🎧', '📻', '🎬', '🎨', '🎭', '🎪', '🎰',
  '🚗', '🚕', '🚙', '🚌', '🚎', '🏎️', '🚓', '🚑',
  '🚒', '🚐', '🛻', '🚚', '🚛', '🚜', '✈️', '🚀',
] as const

/**
 * Convert SAS indices to emoji string
 * @param indices - Array of 7 byte values (0-255)
 * @returns String of 7 emojis
 */
export function indicesToEmojis(indices: readonly number[]): string {
  if (indices.length !== 7) {
    throw new Error('SAS indices must be exactly 7 bytes')
  }
  return indices.map((i) => SAS_EMOJIS[i]).join('')
}

/**
 * Generate SAS from device public keys and nonce
 *
 * ADR-008 algorithm:
 * 1. Concatenate: identity_signing_pk || device_signing_pk || device_ecdh_pk || client_nonce
 * 2. BLAKE3 hash
 * 3. Take first 7 bytes (56 bits)
 * 4. Convert each byte to emoji
 *
 * @param identitySigningPk - User's identity signing public key (32 bytes)
 * @param deviceSigningPk - New device's signing public key (32 bytes)
 * @param deviceEcdhPk - New device's ECDH public key (32 bytes)
 * @param clientNonce - Client-generated nonce (16 bytes)
 * @returns Array of 7 emoji indices (0-255)
 */
export function generateSasIndices(
  identitySigningPk: Uint8Array,
  deviceSigningPk: Uint8Array,
  deviceEcdhPk: Uint8Array,
  clientNonce: Uint8Array
): number[] {
  // Validate input lengths
  if (identitySigningPk.length !== 32) {
    throw new Error('Identity signing public key must be 32 bytes')
  }
  if (deviceSigningPk.length !== 32) {
    throw new Error('Device signing public key must be 32 bytes')
  }
  if (deviceEcdhPk.length !== 32) {
    throw new Error('Device ECDH public key must be 32 bytes')
  }
  if (clientNonce.length !== 16) {
    throw new Error('Client nonce must be 16 bytes')
  }

  // Concatenate all inputs
  const message = new Uint8Array(32 + 32 + 32 + 16)
  message.set(identitySigningPk, 0)
  message.set(deviceSigningPk, 32)
  message.set(deviceEcdhPk, 64)
  message.set(clientNonce, 96)

  // BLAKE3 hash
  const hash = blake3(message)

  // Take first 7 bytes as indices
  return Array.from(hash.slice(0, 7))
}

/**
 * Generate SAS emoji string from device public keys and nonce
 *
 * @param identitySigningPk - User's identity signing public key (32 bytes)
 * @param deviceSigningPk - New device's signing public key (32 bytes)
 * @param deviceEcdhPk - New device's ECDH public key (32 bytes)
 * @param clientNonce - Client-generated nonce (16 bytes)
 * @returns String of 7 emojis
 */
export function generateSasEmojis(
  identitySigningPk: Uint8Array,
  deviceSigningPk: Uint8Array,
  deviceEcdhPk: Uint8Array,
  clientNonce: Uint8Array
): string {
  const indices = generateSasIndices(identitySigningPk, deviceSigningPk, deviceEcdhPk, clientNonce)
  return indicesToEmojis(indices)
}

